python/shotgun_model/data_handler_find.py

```diff
@@ -0,0 +1,384 @@
+                **find_kwargs
+            )
+
+        return request_id
+
+    @log_timing
+    def update_data(self, sg_data):
+        """
+        The counterpart to :meth:`generate_data_request`. When the data
+        request has been carried out, this method should be called by the calling
+        class and the data payload from Shotgun should be provided via the
+        sg_data parameter.
+
+        The shotgun find data is compared against the existing tree and
+        a list of differences is returned, indicating which nodes were
+        added, deleted and modified, on the following form::
+
+            [
+             {
+                "data": ShotgunDataItem instance,
+                "mode": self.UPDATED|ADDED|DELETED
+             },
+             {
+                "data": ShotgunDataItem instance,
+                "mode": self.UPDATED|ADDED|DELETED
+             },
+             ...
+            ]
+
+        :param sg_data: list, resulting from a Shotgun find query
+        :returns: list of updates. see above
+        :raises: :class:`ShotgunModelDataError` if no cache is loaded into memory
+        """
+        self._log_debug("Updating %s with %s shotgun records." % (self, len(sg_data)))
+        self._log_debug("Hierarchy: %s" % self.__hierarchy)
+
+        if self._cache is None:
+            raise ShotgunModelDataError("No data currently loaded in memory!")
+
+        if len(self._cache[self.CACHE_CHILDREN]) == 0:
+            self._log_debug("In-memory cache is empty.")
+
+        # ensure the data is clean
+        self._log_debug("sanitizing data...")
+        sg_data = self._sg_clean_data(sg_data)
+        self._log_debug("...done!")
+
+        self._log_debug("Generating new tree in memory...")
+
+        # create a brand new tree rather than trying to be clever
+        # about how we cull intermediate nodes for deleted items
+        diff_list = []
+        num_adds = 0
+        num_deletes = 0
+        num_modifications = 0
+
+        new_cache = self._create_clear_cache()
+
+        # analyze the incoming shotgun data
+        for sg_item in sg_data:
+
+            sub_tree = new_cache
+            # maintain a hierarchy of unique ids
+
+            # Create items by drilling down the hierarchy
+            for field_name in self.__hierarchy:
+
+                on_leaf_level = (self.__hierarchy[-1] == field_name)
+
+                if not on_leaf_level:
+                    # get the parent uid or None if we are at the root level
+                    parent_uid = sub_tree.get(self.UID)
+                    # generate path for this item
+                    unique_field_value = self.__generate_unique_key(parent_uid, field_name, sg_item)
+                else:
+                    # on the leaf level, use the entity id as the unique key
+                    unique_field_value = sg_item["id"]
+
+                # two distinct cases for leaves and non-leaves
+                if on_leaf_level:
```

Comment:
this code is already fairly complex. i am worried we will make it harder to read if we do what you suggest.